### PR TITLE
SDL-0207 RPC message protection (Revision)

### PR DESCRIPTION
**Pull request by yang1070**
This proposal proposes a design to protect (enable encryption for) RPC messages transmitted between a mobile application and the SDL.

### CODE REVIEW

**Comment by yang1070** [comment]
@jordynmackool this pr is ready for review.

File: proposals/0207-rpc-message-protection.md [hunk 1]
```diff
@@ -171,51 +207,17 @@ Multiple function groups can include the same RPC; each group has its own flag f
...
 
 ### SDL server updates:
-The SDL server may support the new parameter in the policy table. 
+
+The SDL server need to support the new parameters in the policy table.
+* Database and API modifications to support the new `encryption_required` attribute of Functional Groups.
+* UI addition of a checkbox or toggle-switch to enable/disable the state of `encryption_required` for a Functional Group.
+* Database and API modifications to support the new `encryption_required` app_policy attribute.
+* New `encryption_required` server environment variable to indicate whether auto-approved applications should automatically have their `encryption_required` app_policy value set to `true` or `false` (default `false`)
+* UI addition of a checkbox or toggle-switch to enable/disable an individual application version's `encryption_required` app_policy value.
```
Comment (jordynmackool) [inline proposals/0207-rpc-message-protection.md]: Update to:
UI addition of a checkbox or toggle-switch to enable/disable an individual application **version** `encryption_required` app_policy value.

File: proposals/0207-rpc-message-protection.md [hunk 2]
```diff
@@ -37,20 +37,30 @@ A mobile application only shall adopt a security library from an OEM if it uses
 
 
 ### Mobile API change
-We add a new Boolean parameter to the existing `PermissionItem` data type. A list of `PermissionItem` will be received by the app via `OnPermissionsChange` notification, so that the app side knows which RPCs need protection. Please note, to follow the naming conventions, we use `needProtection` in RPC messages and use `need_protection` in policy configuration. However, they shall refer to the same thing. 
+We add two new Boolean parameters with the same name `requireEncryption`. One for app level is added to `OnPermissionsChange` and the other for RPC level is added to the  `PermissionItem` data type. A list of `PermissionItem` will be received by the app via `OnPermissionsChange` notification, so that the app side knows whether the app need encryption or not and which RPCs need encryption. Please note, to follow the naming conventions, we use `requireEncryption` in RPC messages and use `encryption_required` in policy configuration. However, they shall refer to the same thing.
```
[the remark this answers is not in the document]
Comment (yang1070) [reply proposals/0207-rpc-message-protection.md]: updated

File: proposals/0207-rpc-message-protection.md [hunk 3]
```diff
@@ -60,17 +70,11 @@ We also add a new result code. SDL sends this code to a mobile app when it recei
 </enum>
 ```
 
-In this proposal, we say a RPC message needs encryption/protection if the RPC has `needProtection` with value `true` in the `PermissionItem`. We say a RPC message does not need encryption/protection if the RPC has `needProtection` with value `false` or the `needProtection` does not exist in the `PermissionItem`. 
+In this proposal, we say a RPC message needs encryption/protection if the app does not have `requireEncryption`=`false` in the `OnPermissionsChange` and the RPC has `requireEncryption`=`true` in the `PermissionItem`. We say a RPC message does not need encryption/protection if the app has `requireEncryption`=`false` in the `OnPermissionsChange` or the app does not have `requireEncryption`=`false` in the `OnPermissionsChange` but the RPC has `requireEncryption`=`false` or the `requireEncryption` does not exist in the `PermissionItem`.
```
[the remark this answers is not in the document]
Comment (yang1070) [reply proposals/0207-rpc-message-protection.md]: update to

In this proposal, RPC messages need encryption/protection if the app does not have `requireEncryption`=`false` in the `OnPermissionsChange` and the RPC has `requireEncryption`=`true` in the `PermissionItem`. The RPC message does not need encryption/protection if the app has `requireEncryption`=`false` in the `OnPermissionsChange` or the app does not have `requireEncryption`=`false` in the `OnPermissionsChange` and either the RPC has `requireEncryption`=`false` or the `requireEncryption` does not exist in the `PermissionItem`.

File: proposals/0207-rpc-message-protection.md [hunk 4]
```diff
@@ -94,12 +98,12 @@ Here we list existing related RPC and data types for completeness of understandi
...
-- 3. Sometime in between. For example, when the driver activates the app and HMI brings the app to foreground. Because it may take some time to get RPC service encryption enabled for the first time, and there is no overhead after that, we recommend the proxy enable encryption after the app is activated (SDL proxy receives `OnHMIStatus` notification with `hmiLevel=FULL\LIMITED\BACKGROUND`, but not `NONE`) and there is at least one RPC that needs protection.
+- 3. Sometime in between (Proposed option). For example, when the driver activates the app and HMI brings the app to foreground. Because it may take some time to get RPC service encryption enabled for the first time, and there is no overhead after that, we recommend the proxy enable encryption after the app is activated (SDL proxy receives `OnHMIStatus` notification with `hmiLevel=FULL\LIMITED\BACKGROUND`, but not `NONE`) and there is at least one RPC that needs protection.
```
[the remark this answers is not in the document]
Comment (yang1070) [reply proposals/0207-rpc-message-protection.md]: updated

File: proposals/0207-rpc-message-protection.md [hunk 5]
```diff
@@ -94,12 +98,12 @@ Here we list existing related RPC and data types for completeness of understandi
...
 
-When the mobile proxy receives `OnPermissionsChange` message, it shall remember the list of RPCs that need encryption for later use.
+When the mobile proxy receives `OnPermissionsChange` message, it shall remember the flag for the app and the list of RPCs that need encryption if the app's flag is true for later use.
```
[the remark this answers is not in the document]
Comment (yang1070) [reply proposals/0207-rpc-message-protection.md]: updated

File: proposals/0207-rpc-message-protection.md [hunk 6]
```diff
@@ -123,11 +127,39 @@ After the encryption of RPC service 7 is enabled (encryption is available), SDL
 ### Policy updates:
 
 #### Proposed solution
-Add an optional Boolean flag `need_protection` to a function group to indicate whether all the RPCs within this function group need protection not. For example, 
+Add an optional Boolean flag `encryption_required` to each app within `app_policies` to indicate whether the app require encryption or not. SDL core shall translate this flag in policy to `requireEncryption` in the OnPermissionsChange.
```
[the remark this answers is not in the document]
Comment (yang1070) [reply proposals/0207-rpc-message-protection.md]: updated

File: proposals/0207-rpc-message-protection.md [hunk 7]
```diff
@@ -140,29 +172,33 @@ Add an optional Boolean flag `need_protection` to a function group to indicate w
...
 
+Note that even the flag is defined in the function group level in the Policy Table, SDL core still needs to cascade this flag from the function group level to each RPC level because there is no concept of `function group` in `OnPermissionsChange` notification. 
 
-If `need_protection`=`true` for a function group in the Policy Table, all the RPCs within that function group must be sent/received in an encryption enabled SDL service. This means the app has been authenticated via TLS handshake and RPC request and response messages are encrypted. If `need_protection`=`false` or `need_protection` does not exist for a function group in the Policy Table, the RPC messages of that function group shall not be encrypted and can be transmitted in both encryption enabled and disabled SDL services.
+If `encryption_required`=`false` in the app level in `app_policies`, sdl core and mobile proxy shall not enable rpc encryption regardless of the value of `encryption_required` in function group level. SDL core does not need to check the flag in function group level. Mobile proxy does not need to check the flag in RPC level.
```
[the remark this answers is not in the document]
Comment (yang1070) [reply proposals/0207-rpc-message-protection.md]: updated

File: proposals/0207-rpc-message-protection.md [hunk 8]
```diff
@@ -171,51 +207,17 @@ Multiple function groups can include the same RPC; each group has its own flag f
...
-    ...
-}
-```
-
-In this alternative solution, OEMs are forced to choose either no encryption at all or almost all messages get encrypted. OEMs lose the flexibility to configure which function group of RPCs need encryption. Not all OEMs are ready to encrypt all RPC messages for a variety of reasons. At the same time, there is a need to encrypt remote control related RPCs that have a high risk for the new apps on the existing hardware. This proposal proposes the solution of encrypting select RPCs rather than encrypting all RPCs.
 
 ### SDL server updates:
-The SDL server may support the new parameter in the policy table. 
+
+The SDL server need to support the new parameters in the policy table.
```
[the remark this answers is not in the document]
Comment (yang1070) [reply proposals/0207-rpc-message-protection.md]: updated

File: proposals/0207-rpc-message-protection.md (hunk 1 above)
Comment (yang1070) [reply proposals/0207-rpc-message-protection.md]: updated

**Comment by yang1070** [comment]
@jordynmackool this pr is ready for re-review.

File: proposals/0207-rpc-message-protection.md [hunk 9]
```diff
@@ -70,7 +70,7 @@ We also add a new result code. SDL sends this code to a mobile app when it recei
 </enum>
 ```
 
-In this proposal, we say a RPC message needs encryption/protection if the app does not have `requireEncryption`=`false` in the `OnPermissionsChange` and the RPC has `requireEncryption`=`true` in the `PermissionItem`. We say a RPC message does not need encryption/protection if the app has `requireEncryption`=`false` in the `OnPermissionsChange` or the app does not have `requireEncryption`=`false` in the `OnPermissionsChange` but the RPC has `requireEncryption`=`false` or the `requireEncryption` does not exist in the `PermissionItem`. 
+In this proposal, RPC messages need encryption/protection if the app does not have `requireEncryption`=`false` in the `OnPermissionsChange` and the RPC has `requireEncryption`=`true` in the `PermissionItem`. The RPC message does not need encryption/protection if the app has `requireEncryption`=`false` in the `OnPermissionsChange` or the app does not have `requireEncryption`=`false` in the `OnPermissionsChange` and either the RPC has `requireEncryption`=`false` or the `requireEncryption` does not exist in the `PermissionItem`.
```
Comment (jordynmackool) [inline proposals/0207-rpc-message-protection.md]: Please update to this format as it is very hard to follow at its current state. 

In this proposal, RPC messages need encryption/protection if the app does not have `requireEncryption`=`false` in the `OnPermissionsChange` and the RPC has `requireEncryption`=`true` in the `PermissionItem`. 
The RPC message does not need encryption/protection if:
-  The app has`requireEncryption`=`false` in the `OnPermissionsChange`.
- The app does not have `requireEncryption`=`false` in the `OnPermissionsChange`.
- Either **of** the **RPCs** **have** `requireEncryption`=`false`.
- The`requireEncryption` does not exist in the `PermissionItem`.

File: proposals/0207-rpc-message-protection.md [hunk 10]
```diff
@@ -127,7 +127,7 @@ After the encryption of RPC service 7 is enabled (encryption is available), SDL
 ### Policy updates:
 
 #### Proposed solution
-Add an optional Boolean flag `encryption_required` to each app within `app_policies` to indicate whether the app require encryption or not. SDL core shall translate this flag in policy to `requireEncryption` in the OnPermissionsChange.
+Add an optional Boolean flag `encryption_required` to each app within `app_policies` to indicate whether the app requires encryption or not. SDL core shall translate this flag in tje policy to `requireEncryption` in the OnPermissionsChange.
```
Comment (jordynmackool) [inline proposals/0207-rpc-message-protection.md]: please update `tje` to `the`

File: proposals/0207-rpc-message-protection.md (hunk 9 above)
Comment (yang1070) [reply proposals/0207-rpc-message-protection.md]: In this proposal, RPC messages need encryption/protection if the app does not have `requireEncryption`=`false` in the `OnPermissionsChange` and the RPC has `requireEncryption`=`true` in the `PermissionItem`. 

The RPC message does not need encryption/protection if:
- The app has `requireEncryption`=`false` in the `OnPermissionsChange`. 
- The app does not have `requireEncryption`=`false` in the `OnPermissionsChange` and either the RPC has `requireEncryption`=`false` or the `requireEncryption` does not exist in the `PermissionItem`.

File: proposals/0207-rpc-message-protection.md (hunk 10 above)
Comment (yang1070) [reply proposals/0207-rpc-message-protection.md]: update

**Comment by yang1070** [comment]
@jordynmackool this pr is ready for re-review.

**Comment by yang1070** [comment]
@jordynmackool I remember the proposal was accepted with revision. Does it need to be marked as "review ready"?

**Comment by yang1070** [comment]
@jordynmackool This PR is an update/revision according to https://github.com/smartdevicelink/sdl_evolution/issues/634, I think you might misplace the tag "review ready"
